### PR TITLE
[v0.6.0] coroutines

### PR DESCRIPTION
**Pull request by modrzew**
This PR moves away from using 1 thread per worker in favor of using Python coroutines. It's **still in progress**, but has been already pretty stable for past few days, so I think about merging it back to master.

**Does it work?** It sure does, I've been running only this version over past few days.
**Is it faster than threaded version?** Performance-wise it's the same, if not slightly worse.
**Is it better than threaded version?** From user's perspective? It may be, as you can configure how many threads you want to use, so you may need less resources. From my perspective as developer? _Way better_. There's much more control over flow than in the threaded version.

And when I say "coroutines", I also mean that...

---
# Python 2.7 is no longer supported.
## Python 3.5 is supported.

Python 3.4 could have been supported, but I decided to go with new fancy syntax that includes `async def` and `await`. Sorry!

---

Alright, so you'll need two new config options: `COMPUTE_THREADS` and `NETWORK_THREADS`. All operations are done in the single thread _except_ for API requests and obtaining cell IDs, which are done in separate threads. My recommended values:
- `COMPUTE_THREADS`: amount of workers / 10
- `NETWORK_THREADS`: amount of workers / 2

That gives bigger control over how many threads are used. I found out that 180 workers can live with ~100 threads used for networking. Of course you can still use as many network threads as workers.

Moreover, besides coroutines, only one thread (DatabaseProcessor) talks to database at the single time. This greatly reduces resources necessary on communicating with it, decreases chances of duplicated entries, as well as makes all SQLite-related errors obsolete. (At least in theory.) It also paves way for other integration, such as webhooks. Workers now focus solely on scanning the map and putting raw results in the queue.

Also there's new enhanced status window that is better suited for displaying lots of workers. It displays more information in concise way, and shows worker status only if something bad happened. And other fixes here and there.

**Give it a try and tell me how it works!** As I said, I have been running and polishing this version for past few days, and generally I'm happy with it.
### Todo
- [ ] investigate and improve exception handling in coroutines (right now they fail silently, which is quite bad)

### CODE REVIEW

**Comment by YonderGod** [comment]
cool!

running with no problems so far.

**Comment by Aiyubi** [comment]
things I noticed:
* you recommend values for threads but hardcode a 10 - why not just app the recommendations rounded up the config.py.example?
* the status bar is .... weird and doesn't really show much. Maybe do a linebreak every config.GRID[1] elements and mirror the lines? This way they are ordered in the same way the workers are ordered. Might also look weird. Currently it is so irritating cause you dont see any progress on pokemonfindings.
* it added all (330) datapoints 24 days in the future at the same normalized timestamp. It does not find more pokemon ...

**Comment by Aiyubi** [comment]
okay. Got some more testing done with 1 hour of runtime:
* missing pokemonprogress makes it hard to see if everything is running
* no data gets written to the database while discoveringprocess. It only starts writing to the database after first Keyboardinterrupt. The data that gets written is not as much as it should be. It seems to only be the first 10ish minutes of the timewindow.
* need to do 3 keyinterrupts: first to stop discovering. Second to stop main thread(?), third to interrupt interrupthandling

**Comment by modrzew** [comment]
@Aiyubi thanks, I added bunch of stuff in the most recent commits - you can see diff here: https://github.com/modrzew/pokeminer/compare/153a030...179bb55

**Comment by DavidSelem** [comment]
When trying to access to my PC from outside, mysql process goes crazy, upto 90% CPU

**Comment by Aiyubi** [comment]
Looks now way better and works. Only problem is I'm running into a segmentation fault now. Mamory access one. Happened each time after ~1700 pokemon (tested 3 times so far).

**Comment by vDrag0n** [comment]
It could just be my stack, but after awhile many of my accounts have exceptions in the status. 

`[2016-08-16 22:46:13,763][    INFO][worker-24] Restarting
[2016-08-16 22:46:13,764][    INFO][requests.packages.urllib3.connectionpool] Starting new HTTPS connection (1): sso.pokemon.com
[2016-08-16 22:46:13,765][   ERROR][worker-49] A wild exception appeared!
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 142, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/util/connection.py", line 67, in create_connection
    for res in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 578, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 351, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 814, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 254, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 151, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
requests.packages.urllib3.exceptions.NewConnectionError: <requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10c2e2c88>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/adapters.py", line 403, in send
    timeout=timeout
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 623, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py", line 281, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
requests.packages.urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='sso.pokemon.com', port=443): Max retries exceeded with url: /sso/login?service=https%3A%2F%2Fsso.pokemon.com%2Fsso%2Foauth2.0%2FcallbackAuthorize (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10c2e2c88>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "worker.py", line 148, in login
    provider=config.ACCOUNTS[self.worker_no][2],
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "worker.py", line 128, in inner
    result = method(*args, **kwargs)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/pgoapi.py", line 169, in login
    self.set_authentication(provider, username=username, password=password)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/pgoapi.py", line 89, in set_authentication
    self._auth_provider.user_login(username, password)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/auth_ptc.py", line 63, in user_login
    r = self._session.get(self.PTC_LOGIN_URL, headers=head)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 487, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 585, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/adapters.py", line 467, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='sso.pokemon.com', port=443): Max retries exceeded with url: /sso/login?service=https%3A%2F%2Fsso.pokemon.com%2Fsso%2Foauth2.0%2FcallbackAuthorize (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10c2e2c88>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))
[2016-08-16 22:46:13,766][    INFO][worker-49] Restarting
[2016-08-16 22:46:13,766][   ERROR][worker-73] A wild exception appeared!
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 142, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/util/connection.py", line 67, in create_connection
    for res in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/socket.py", line 732, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 578, in urlopen
    chunked=chunked)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 351, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 814, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 254, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 151, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
requests.packages.urllib3.exceptions.NewConnectionError: <requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10a409240>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/requests/adapters.py", line 403, in send
    timeout=timeout
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 623, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/local/lib/python3.5/site-packages/requests/packages/urllib3/util/retry.py", line 281, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
requests.packages.urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='sso.pokemon.com', port=443): Max retries exceeded with url: /sso/login?service=https%3A%2F%2Fsso.pokemon.com%2Fsso%2Foauth2.0%2FcallbackAuthorize (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10a409240>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "worker.py", line 148, in login
    provider=config.ACCOUNTS[self.worker_no][2],
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/futures.py", line 361, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/tasks.py", line 296, in _wakeup
    future.result()
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
  File "/usr/local/Cellar/python3/3.5.2_1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "worker.py", line 128, in inner
    result = method(*args, **kwargs)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/pgoapi.py", line 169, in login
    self.set_authentication(provider, username=username, password=password)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/pgoapi.py", line 89, in set_authentication
    self._auth_provider.user_login(username, password)
  File "/Users/Apple/pokeminer-coroutines/src/pgoapi/pgoapi/auth_ptc.py", line 63, in user_login
    r = self._session.get(self.PTC_LOGIN_URL, headers=head)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 487, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 475, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/sessions.py", line 585, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/requests/adapters.py", line 467, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='sso.pokemon.com', port=443): Max retries exceeded with url: /sso/login?service=https%3A%2F%2Fsso.pokemon.com%2Fsso%2Foauth2.0%2FcallbackAuthorize (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x10a409240>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known',))`

**Comment by modrzew** [comment]
Nothing strange with your internet connection? I've been running it for 2 days previously, and now it's been running for 14 hours straight without any hiccups - neither segfault nor network issues.

**Comment by vDrag0n** [comment]
I run scans with the older version and no problems for hours. 

When running the coroutine, the accounts will continuously get the exception.
after i restart the scan, no errors pop up.  

Im wondering if theres a limit to ssl connections that can be set somewhere for python 3.5 on OSX

**Comment by snicker** [comment]
hey @modrzew can I run this alongside the current master branch against the same database? I'm using MySQL, I just want to make sure there aren't any schema changes. Interested in helping test it out, I'll split my accounts and city in half and see which version is more stable.

**Comment by Noctem** [comment]
This works pretty well for me, though I did experience a segfault after running on OS X for over 36 hours.

**Comment by fourbytes** [comment]
Has been working perfectly for me for the last couple of days.

File: worker.py [hunk 1]
```diff
@@ -621,6 +624,8 @@ def get_status_message(self):
             'Pokemon found count (10s interval):',
             ' '.join(self.things_count),
             '',
+            'Workers without sightings so far:',
+            ', '.join(w for w in self.workers.values() if w.total_seen == 0),
```
Comment (Noctem) [inline worker.py]: `TypeError: sequence item 0: expected str instance, Slave found`

File: worker.py [hunk 2]
```diff
@@ -688,6 +700,7 @@ def exception_handler(loop, context):
         loop.run_forever()
     except KeyboardInterrupt:
         print('Exiting, please wait until all tasks finish')
-        overseer.kill()
-        loop.run_until_complete(asyncio.gather(*asyncio.Task.all_tasks()))
+        overseer.kill()  # also cancels all workers' futures
+        all_futures = [w.future for w in overseer.workers.values()]
+        loop.run_until_complete(asyncio.gather(*all_futures))
```
Comment (Noctem) [inline worker.py]: `TypeError: A Future, a coroutine or an awaitable is required`

File: worker.py [hunk 3]
```diff
@@ -388,13 +698,37 @@ def parse_args():
...
+    try:
+        loop.run_forever()
+    except KeyboardInterrupt:
+        print('Exiting, please wait until all tasks finish')
+        overseer.kill()  # also cancels all workers' futures
+        all_futures = [
+            w.future for w in overseer.workers.values()
+            if w.future is not None
+        ]
+        loop.run_until_complete(asyncio.gather(*all_futures))
```
Comment (Noctem) [inline worker.py]: `TypeError: A Future, a coroutine or an awaitable is required`

**Comment by Aiyubi** [comment]
* did some further testing on fresh setups (new computer + database). still running in the coredump at some point (sometimes takes half a day)
* Getting the following on my slower computer and can not find a solution:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/threading.py", line 923, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.5/threading.py", line 871, in run
    self._target(*self._args, **self._kwargs)
  File "worker.py", line 460, in check
    self.start_worker(worker_no)
  File "worker.py", line 431, in start_worker
    asyncio.run_coroutine_threadsafe(worker.run(), self.loop)
AttributeError: module 'asyncio' has no attribute 'run_coroutine_threadsafe'
```
And in case the question comes up:
```
pip3.5 show asyncio
---
Metadata-Version: 2.0
Name: asyncio
Version: 3.4.3
Summary: reference implementation of PEP 3156
Home-page: http://www.python.org/dev/peps/pep-3156/
Author: UNKNOWN
Author-email: UNKNOWN
Installer: pip
License: UNKNOWN
Location: /usr/local/lib/python3.5/site-packages
Requires:
Classifiers:
  Programming Language :: Python
  Programming Language :: Python :: 3
  Programming Language :: Python :: 3.3

```

**Comment by modrzew** [comment]
IIRC `run_coroutine_threadsafe` came to life in Python 3.5.1, so the quickest solution is to upgrade. I don't know if there is a better way of starting worker (in event loop) from overseer's thread.

**Comment by Spacehug** [comment]
@modrzew I fucking love you for py3

**Pull request closed without merging**